<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>The robust generalization performance of the STAVER algorithm across diverse and large-scale DIA datasets &mdash; STAVER 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/documentation_options.js?v=01f34227"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The robustness and broad applicability of the STAVER algorithm for disease diagnosis and classification." href="PanCancer_949_cellline_models.html" />
    <link rel="prev" title="Run the STAVER workflow" href="staver_pipeline_cli.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            STAVER
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dataset.html">Example datasets</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="staver_pipeline_cli.html">Run the STAVER workflow</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">The robust generalization performance of the STAVER algorithm across diverse and large-scale DIA datasets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Metadata">Metadata</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Tissue_type">Tissue_type</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Cancer_type">Cancer_type</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Cancer_subtype">Cancer_subtype</a></li>
<li class="toctree-l3"><a class="reference internal" href="#The-Sankey-diagram-delineates-the-relationships">The Sankey diagram delineates the relationships</a></li>
<li class="toctree-l3"><a class="reference internal" href="#The-reproducibility-and-reliability-of-the-STAVER-processed-data">The reproducibility and reliability of the STAVER-processed data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#HEK293T-Spearman-correlation-analysis">HEK293T Spearman correlation analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-correlation-heatmap-of-raw-data">The correlation heatmap of raw data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-correlation-heatmap-of-STAVER-processed-data">The correlation heatmap of STAVER-processed data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-IQR-of-HEK293T-spearman-correlation-matrix-in-rawdata">The IQR of HEK293T spearman correlation matrix in rawdata</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-IQR-of-HEK293T-spearman-correlation-matrix-in-SATVER-processed-data">The IQR of HEK293T spearman correlation matrix in SATVER-processed data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-identification-protein-numbers-of-the-raw-data-and-STAVER-data">The identification protein numbers of the raw data and STAVER data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-Coefficient-of-Variation-(CVs)-of-the-raw-data-and-STAVER-data">The Coefficient of Variation (CVs) of the raw data and STAVER data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#The-advantages-of-the-STAVER-algorithm-to-uncover-inherent-biological-differences">The advantages of the STAVER algorithm to uncover inherent biological differences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#The-UMAP-analysis-of-diverse-1242-cancer-cell-line-samples">The UMAP analysis of diverse 1242 cancer cell line samples</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-UMAP-analysis-of-the-Rawdata">The UMAP analysis of the Rawdata</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-UMAP-analysis-of-STAVER-processed-data">The UMAP analysis of STAVER-processed data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Cancer-specific-proteins">Cancer specific proteins</a></li>
<li class="toctree-l4"><a class="reference internal" href="#CTSE-overexpression-in-Gastric,-pancreatic,-and-colorectal-cancer-of-HPA-dataset">CTSE overexpression in Gastric, pancreatic, and colorectal cancer of HPA dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#GPA33-overexpression-in-colorectal-cancer-and-Gastric-cancer-of-HPA-dataset">GPA33 overexpression in colorectal cancer and Gastric cancer of HPA dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ADGRF1-overexpression-in-pancreatic-cancer-of-HPA-dataset">ADGRF1 overexpression in pancreatic cancer of HPA dataset</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#The-reproducibility-of-the-STAVER-algorithm-in-identifying-the-previously-reported-tumor-biomarkers">The reproducibility of the STAVER algorithm in identifying the previously reported tumor biomarkers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#The-robustness-and-broad-applicability-of-the-STAVER-algorithm-for-disease-diagnosis-and-classification.">The robustness and broad applicability of the STAVER algorithm for disease diagnosis and classification.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="PanCancer_949_cellline_models.html">The robustness and broad applicability of the STAVER algorithm for disease diagnosis and classification.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="release.html">Release notes</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">STAVER</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="dataset.html">Example datasets</a></li>
      <li class="breadcrumb-item active">The robust generalization performance of the STAVER algorithm across diverse and large-scale DIA datasets</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/PanCancer_949_celline_analysis.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="The-robust-generalization-performance-of-the-STAVER-algorithm-across-diverse-and-large-scale-DIA-datasets">
<h1>The robust generalization performance of the STAVER algorithm across diverse and large-scale DIA datasets<a class="headerlink" href="#The-robust-generalization-performance-of-the-STAVER-algorithm-across-diverse-and-large-scale-DIA-datasets" title="Link to this heading">¶</a></h1>
<p>To further validate the reliability of STAVER’s results and underscore the robustness and inherent advantages of the STAVER algorithm, we applied STAVER to a much more diverse and larger-scale DIA dataset from the “ProCan-DepMap-Sanger project” (<a class="reference external" href="https://doi:10.1016/j.ccell.2022.06.010">Gonçalves, et al. 2022, Cancer Cell</a>). A total of 1,326 samples were included for analysis, including 84 samples of HEK293T cell lines used for quality control and 1,242 cancer cell samples derived from 9
typical cancer types (colorectal, SCLC, kidney, gastric, pancreatic, bladder, NSCLC, glioma, and hepatocellular cancer) <strong>(Figure RL2A-RL2B and Figure RL5A-RL5B)</strong>.</p>
<p>We performed a comprehensive comparative analysis of the original data without the STAVER processing and the STAVER-processed data, with the main results focusing on - (I) the reproducibility and reliability of the STAVER-processed data, - (II) the robustness of the STAVER algorithm to uncover inherent biological differences, - (III) the reproducibility of the STAVER algorithm in identifying the previously reported tumor biomarkers, and - (IV) the robustness and broad applicability of the STAVER
algorithm for disease diagnosis and classification.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">plotly.express</span> <span class="k">as</span> <span class="nn">px</span>
<span class="kn">import</span> <span class="nn">plotly.io</span> <span class="k">as</span> <span class="nn">pio</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">colorsys</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="c1"># python matplotlib export editable PDF</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;pdf.fonttype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">42</span>
<span class="c1"># mpl.rcParams[&#39;figure.dpi&#39;]= 150</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">plotly.offline</span> <span class="kn">import</span> <span class="n">init_notebook_mode</span><span class="p">,</span> <span class="n">iplot</span>
<span class="n">init_notebook_mode</span><span class="p">(</span><span class="n">connected</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<script type="text/javascript">
window.PlotlyConfig = {MathJaxConfig: 'local'};
if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}
if (typeof require !== 'undefined') {
require.undef("plotly");
requirejs.config({
    paths: {
        'plotly': ['https://cdn.plot.ly/plotly-2.26.0.min']
    }
});
require(['plotly'], function(Plotly) {
    window._Plotly = Plotly;
});
}
</script></div>
</div>
<section id="Metadata">
<h2>Metadata<a class="headerlink" href="#Metadata" title="Link to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/Pan-cancer-cell-lines/subset_949_samples.xlsx&quot;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Automatic_MS_filename</th>
      <th>Batch</th>
      <th>Date</th>
      <th>Instrument</th>
      <th>Cell_line</th>
      <th>SIDM</th>
      <th>Tissue_type</th>
      <th>Cancer_type</th>
      <th>Cancer_subtype</th>
      <th>Project_Identifier</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>519</th>
      <td>191012_b36-t2-8_00di6_00jm7_m06_s_1</td>
      <td>P04</td>
      <td>2019-10-12</td>
      <td>M06</td>
      <td>BFTC-905</td>
      <td>SIDM00989</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Urothelial carcinoma</td>
      <td>SIDM00989;BFTC-905</td>
    </tr>
    <tr>
      <th>521</th>
      <td>191023_b61-t1-1_00dsu_00kp3_m03_s_1</td>
      <td>P04</td>
      <td>2019-10-23</td>
      <td>M03</td>
      <td>BFTC-905</td>
      <td>SIDM00989</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Urothelial carcinoma</td>
      <td>SIDM00989;BFTC-905</td>
    </tr>
    <tr>
      <th>522</th>
      <td>191026_b36-t3-8_00di6_00kt6_m04_s_1</td>
      <td>P04</td>
      <td>2019-10-26</td>
      <td>M04</td>
      <td>BFTC-905</td>
      <td>SIDM00989</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Urothelial carcinoma</td>
      <td>SIDM00989;BFTC-905</td>
    </tr>
    <tr>
      <th>523</th>
      <td>191026_b61-t2-1_00dsu_00ktw_m06_s_1</td>
      <td>P04</td>
      <td>2019-10-26</td>
      <td>M06</td>
      <td>BFTC-905</td>
      <td>SIDM00989</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Urothelial carcinoma</td>
      <td>SIDM00989;BFTC-905</td>
    </tr>
    <tr>
      <th>524</th>
      <td>191125_b32-t4-8_00dge_00n1d_m03_s_1</td>
      <td>P04</td>
      <td>2019-11-25</td>
      <td>M03</td>
      <td>BFTC-905</td>
      <td>SIDM00989</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Urothelial carcinoma</td>
      <td>SIDM00989;BFTC-905</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6590</th>
      <td>200131_b4-7-t3-1_00q3n_00rtc_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>COR-L95</td>
      <td>SIDM00521</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00521;COR-L95</td>
    </tr>
    <tr>
      <th>6591</th>
      <td>200131_b4-9-t3-1_00q3p_00rte_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H510A</td>
      <td>SIDM00927</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00927;NCI-H510A</td>
    </tr>
    <tr>
      <th>6592</th>
      <td>200131_b4-10-t3-1_00q3q_00rtf_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H2171</td>
      <td>SIDM00733</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00733;NCI-H2171</td>
    </tr>
    <tr>
      <th>6593</th>
      <td>200131_b4-13-t3-1_00q3t_00rti_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H1836</td>
      <td>SIDM00770</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00770;NCI-H1836</td>
    </tr>
    <tr>
      <th>6594</th>
      <td>200201_b3-13-t3-2_00q3d_00rtm_m05_s_1</td>
      <td>P06</td>
      <td>2020-02-01</td>
      <td>M05</td>
      <td>IST-SL1</td>
      <td>SIDM00223</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00223;IST-SL1</td>
    </tr>
  </tbody>
</table>
<p>1242 rows × 10 columns</p>
</div></div>
</div>
</section>
<section id="Tissue_type">
<h2>Tissue_type<a class="headerlink" href="#Tissue_type" title="Link to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Tissue_type_counts</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">Tissue_type_counts</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">,</span> <span class="s1">&#39;Count&#39;</span><span class="p">]</span>
<span class="n">Tissue_type_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Tissue_type</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Lung</td>
      <td>273</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Large Intestine</td>
      <td>271</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Kidney</td>
      <td>181</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Stomach</td>
      <td>153</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Pancreas</td>
      <td>115</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Bladder</td>
      <td>96</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Central Nervous System</td>
      <td>78</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Liver</td>
      <td>75</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">Tissue_type_counts</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Diverse Tissue Type&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">update_traces</span><span class="p">(</span><span class="n">textinfo</span><span class="o">=</span><span class="s1">&#39;label+percent&#39;</span><span class="p">,</span> <span class="n">insidetextorientation</span><span class="o">=</span><span class="s1">&#39;radial&#39;</span><span class="p">)</span>

<span class="c1"># Save figure to PDF</span>
<span class="n">pio</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="s1">&#39;figs/Tissue_type.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>                            <div id="62bd9eef-5089-42a1-bc8c-acbc82c819f9" class="plotly-graph-div" style="height:525px; width:100%;"></div>            <script type="text/javascript">                require(["plotly"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("62bd9eef-5089-42a1-bc8c-acbc82c819f9")) {                    Plotly.newPlot(                        "62bd9eef-5089-42a1-bc8c-acbc82c819f9",                        [{"domain":{"x":[0.0,1.0],"y":[0.0,1.0]},"hovertemplate":"Tissue_type=%{label}\u003cbr\u003eCount=%{value}\u003cextra\u003e\u003c\u002fextra\u003e","labels":["Lung","Large Intestine","Kidney","Stomach","Pancreas","Bladder","Central Nervous System","Liver"],"legendgroup":"","name":"","showlegend":true,"values":[273,271,181,153,115,96,78,75],"type":"pie","insidetextorientation":"radial","textinfo":"label+percent"}],                        {"template":{"data":{"histogram2dcontour":[{"type":"histogram2dcontour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"choropleth":[{"type":"choropleth","colorbar":{"outlinewidth":0,"ticks":""}}],"histogram2d":[{"type":"histogram2d","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmap":[{"type":"heatmap","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmapgl":[{"type":"heatmapgl","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"contourcarpet":[{"type":"contourcarpet","colorbar":{"outlinewidth":0,"ticks":""}}],"contour":[{"type":"contour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"surface":[{"type":"surface","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"mesh3d":[{"type":"mesh3d","colorbar":{"outlinewidth":0,"ticks":""}}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"parcoords":[{"type":"parcoords","line":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolargl":[{"type":"scatterpolargl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"scattergeo":[{"type":"scattergeo","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolar":[{"type":"scatterpolar","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"scattergl":[{"type":"scattergl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatter3d":[{"type":"scatter3d","line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermapbox":[{"type":"scattermapbox","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterternary":[{"type":"scatterternary","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattercarpet":[{"type":"scattercarpet","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"pie":[{"automargin":true,"type":"pie"}]},"layout":{"autotypenumbers":"strict","colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"hovermode":"closest","hoverlabel":{"align":"left"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"bgcolor":"#E5ECF6","angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"ternary":{"bgcolor":"#E5ECF6","aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]]},"xaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"yaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"geo":{"bgcolor":"white","landcolor":"#E5ECF6","subunitcolor":"white","showland":true,"showlakes":true,"lakecolor":"white"},"title":{"x":0.05},"mapbox":{"style":"light"}}},"legend":{"tracegroupgap":0},"title":{"text":"Diverse Tissue Type"}},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('62bd9eef-5089-42a1-bc8c-acbc82c819f9');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                });            </script>        </div></div>
</div>
</section>
<section id="Cancer_type">
<h2>Cancer_type<a class="headerlink" href="#Cancer_type" title="Link to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Cancer_type_counts</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">Cancer_type_counts</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">,</span> <span class="s1">&#39;Count&#39;</span><span class="p">]</span>
<span class="n">Cancer_type_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cancer_type</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Colorectal Carcinoma</td>
      <td>271</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Small Cell Lung Carcinoma</td>
      <td>187</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Kidney Carcinoma</td>
      <td>181</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Gastric Carcinoma</td>
      <td>153</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Pancreatic Carcinoma</td>
      <td>110</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Bladder Carcinoma</td>
      <td>96</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Non-Small Cell Lung Carcinoma</td>
      <td>86</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Glioma</td>
      <td>78</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Hepatocellular Carcinoma</td>
      <td>75</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Other Solid Carcinomas</td>
      <td>5</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">Cancer_type_counts</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Diverse Cancer Type&quot;</span><span class="p">)</span>
<span class="c1"># Update the labels to show both count and percentage</span>
<span class="n">fig</span><span class="o">.</span><span class="n">update_traces</span><span class="p">(</span><span class="n">textinfo</span><span class="o">=</span><span class="s1">&#39;label+percent&#39;</span><span class="p">,</span> <span class="n">insidetextorientation</span><span class="o">=</span><span class="s1">&#39;radial&#39;</span><span class="p">)</span>

<span class="c1"># Save figure to PDF</span>
<span class="n">pio</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="s1">&#39;figs/Cancer_types.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>                            <div id="de7f5c75-8bec-45e6-b848-689a0a0f58eb" class="plotly-graph-div" style="height:525px; width:100%;"></div>            <script type="text/javascript">                require(["plotly"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("de7f5c75-8bec-45e6-b848-689a0a0f58eb")) {                    Plotly.newPlot(                        "de7f5c75-8bec-45e6-b848-689a0a0f58eb",                        [{"domain":{"x":[0.0,1.0],"y":[0.0,1.0]},"hovertemplate":"Cancer_type=%{label}\u003cbr\u003eCount=%{value}\u003cextra\u003e\u003c\u002fextra\u003e","labels":["Colorectal Carcinoma","Small Cell Lung Carcinoma","Kidney Carcinoma","Gastric Carcinoma","Pancreatic Carcinoma","Bladder Carcinoma","Non-Small Cell Lung Carcinoma","Glioma","Hepatocellular Carcinoma","Other Solid Carcinomas"],"legendgroup":"","name":"","showlegend":true,"values":[271,187,181,153,110,96,86,78,75,5],"type":"pie","insidetextorientation":"radial","textinfo":"label+percent"}],                        {"template":{"data":{"histogram2dcontour":[{"type":"histogram2dcontour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"choropleth":[{"type":"choropleth","colorbar":{"outlinewidth":0,"ticks":""}}],"histogram2d":[{"type":"histogram2d","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmap":[{"type":"heatmap","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmapgl":[{"type":"heatmapgl","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"contourcarpet":[{"type":"contourcarpet","colorbar":{"outlinewidth":0,"ticks":""}}],"contour":[{"type":"contour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"surface":[{"type":"surface","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"mesh3d":[{"type":"mesh3d","colorbar":{"outlinewidth":0,"ticks":""}}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"parcoords":[{"type":"parcoords","line":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolargl":[{"type":"scatterpolargl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"scattergeo":[{"type":"scattergeo","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolar":[{"type":"scatterpolar","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"scattergl":[{"type":"scattergl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatter3d":[{"type":"scatter3d","line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermapbox":[{"type":"scattermapbox","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterternary":[{"type":"scatterternary","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattercarpet":[{"type":"scattercarpet","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"pie":[{"automargin":true,"type":"pie"}]},"layout":{"autotypenumbers":"strict","colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"hovermode":"closest","hoverlabel":{"align":"left"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"bgcolor":"#E5ECF6","angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"ternary":{"bgcolor":"#E5ECF6","aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]]},"xaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"yaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"geo":{"bgcolor":"white","landcolor":"#E5ECF6","subunitcolor":"white","showland":true,"showlakes":true,"lakecolor":"white"},"title":{"x":0.05},"mapbox":{"style":"light"}}},"legend":{"tracegroupgap":0},"title":{"text":"Diverse Cancer Type"}},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('de7f5c75-8bec-45e6-b848-689a0a0f58eb');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                });            </script>        </div></div>
</div>
</section>
<section id="Cancer_subtype">
<h2>Cancer_subtype<a class="headerlink" href="#Cancer_subtype" title="Link to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Cancer_subtype</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">Cancer_subtype</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">,</span> <span class="s1">&#39;Count&#39;</span><span class="p">]</span>

<span class="n">Cancer_subtype</span><span class="p">[</span><span class="s2">&quot;Cancer_subtype_modify&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">x</span> <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;</span> <span class="mi">14</span> <span class="k">else</span> <span class="s2">&quot;Others&quot;</span>
    <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">Cancer_subtype</span><span class="p">[</span><span class="s2">&quot;Cancer_subtype&quot;</span><span class="p">],</span> <span class="n">Cancer_subtype</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="p">]</span>

<span class="n">Cancer_subtype</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cancer_subtype</th>
      <th>Count</th>
      <th>Cancer_subtype_modify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Small cell lung carcinoma</td>
      <td>187</td>
      <td>Small cell lung carcinoma</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Colon adenocarcinoma</td>
      <td>147</td>
      <td>Colon adenocarcinoma</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Kidney carcinoma</td>
      <td>107</td>
      <td>Kidney carcinoma</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Bladder carcinoma</td>
      <td>89</td>
      <td>Bladder carcinoma</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Squamous cell lung carcinoma</td>
      <td>86</td>
      <td>Squamous cell lung carcinoma</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Gastric adenocarcinoma</td>
      <td>75</td>
      <td>Gastric adenocarcinoma</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Hepatocellular carcinoma</td>
      <td>72</td>
      <td>Hepatocellular carcinoma</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Low grade glioma</td>
      <td>72</td>
      <td>Low grade glioma</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Clear cell renal cell carcinoma</td>
      <td>64</td>
      <td>Clear cell renal cell carcinoma</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Pancreatic ductal adenocarcinoma</td>
      <td>52</td>
      <td>Pancreatic ductal adenocarcinoma</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Cecum adenocarcinoma</td>
      <td>46</td>
      <td>Cecum adenocarcinoma</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Pancreatic adenocarcinoma</td>
      <td>44</td>
      <td>Pancreatic adenocarcinoma</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Colorectal carcinoma</td>
      <td>43</td>
      <td>Colorectal carcinoma</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Rectal adenocarcinoma</td>
      <td>35</td>
      <td>Rectal adenocarcinoma</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Gastric signet ring cell adenocarcinoma</td>
      <td>29</td>
      <td>Gastric signet ring cell adenocarcinoma</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Gastric tubular adenocarcinoma</td>
      <td>15</td>
      <td>Gastric tubular adenocarcinoma</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Pancreatic carcinoma</td>
      <td>13</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Gastric carcinoma</td>
      <td>12</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Urothelial carcinoma</td>
      <td>7</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Gastric fundus carcinoma</td>
      <td>7</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>20</th>
      <td>Papillary renal cell carcinoma</td>
      <td>6</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>21</th>
      <td>Oligodendroglioma</td>
      <td>6</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>22</th>
      <td>Gastic small cell neuroendocrine carcinoma</td>
      <td>6</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>23</th>
      <td>Gastric small cell carcinoma</td>
      <td>6</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>24</th>
      <td>Pancreatic somatostatinoma</td>
      <td>5</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>25</th>
      <td>Renal pelvis and ureter urothelial carcinoma</td>
      <td>4</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>26</th>
      <td>Hepatoblastoma</td>
      <td>3</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>27</th>
      <td>Gastric choriocarcinoma</td>
      <td>3</td>
      <td>Others</td>
    </tr>
    <tr>
      <th>28</th>
      <td>Pancreatic adenosquamous carcinoma</td>
      <td>1</td>
      <td>Others</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">Cancer_subtype</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Count&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="s1">&#39;Cancer_subtype_modify&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Diverse Cancer subtype&quot;</span><span class="p">)</span>
<span class="c1"># Update the labels to show both count and percentage</span>
<span class="n">fig</span><span class="o">.</span><span class="n">update_traces</span><span class="p">(</span><span class="n">textinfo</span><span class="o">=</span><span class="s1">&#39;label+percent&#39;</span><span class="p">,</span> <span class="n">insidetextorientation</span><span class="o">=</span><span class="s1">&#39;radial&#39;</span><span class="p">)</span>

<span class="c1"># Save figure to PDF</span>
<span class="n">pio</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="s1">&#39;figs/Diverse Cancer subtype modify.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>                            <div id="4743743c-77b7-4b27-8fa4-2fe05a1be638" class="plotly-graph-div" style="height:525px; width:100%;"></div>            <script type="text/javascript">                require(["plotly"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("4743743c-77b7-4b27-8fa4-2fe05a1be638")) {                    Plotly.newPlot(                        "4743743c-77b7-4b27-8fa4-2fe05a1be638",                        [{"domain":{"x":[0.0,1.0],"y":[0.0,1.0]},"hovertemplate":"Cancer_subtype_modify=%{label}\u003cbr\u003eCount=%{value}\u003cextra\u003e\u003c\u002fextra\u003e","labels":["Small cell lung carcinoma","Colon adenocarcinoma","Kidney carcinoma","Bladder carcinoma","Squamous cell lung carcinoma","Gastric adenocarcinoma","Hepatocellular carcinoma","Low grade glioma","Clear cell renal cell carcinoma","Pancreatic ductal adenocarcinoma","Cecum adenocarcinoma","Pancreatic adenocarcinoma","Colorectal carcinoma","Rectal adenocarcinoma","Gastric signet ring cell adenocarcinoma","Gastric tubular adenocarcinoma","Others","Others","Others","Others","Others","Others","Others","Others","Others","Others","Others","Others","Others"],"legendgroup":"","name":"","showlegend":true,"values":[187,147,107,89,86,75,72,72,64,52,46,44,43,35,29,15,13,12,7,7,6,6,6,6,5,4,3,3,1],"type":"pie","insidetextorientation":"radial","textinfo":"label+percent"}],                        {"template":{"data":{"histogram2dcontour":[{"type":"histogram2dcontour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"choropleth":[{"type":"choropleth","colorbar":{"outlinewidth":0,"ticks":""}}],"histogram2d":[{"type":"histogram2d","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmap":[{"type":"heatmap","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmapgl":[{"type":"heatmapgl","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"contourcarpet":[{"type":"contourcarpet","colorbar":{"outlinewidth":0,"ticks":""}}],"contour":[{"type":"contour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"surface":[{"type":"surface","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"mesh3d":[{"type":"mesh3d","colorbar":{"outlinewidth":0,"ticks":""}}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"parcoords":[{"type":"parcoords","line":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolargl":[{"type":"scatterpolargl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"scattergeo":[{"type":"scattergeo","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolar":[{"type":"scatterpolar","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"scattergl":[{"type":"scattergl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatter3d":[{"type":"scatter3d","line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermapbox":[{"type":"scattermapbox","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterternary":[{"type":"scatterternary","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattercarpet":[{"type":"scattercarpet","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"pie":[{"automargin":true,"type":"pie"}]},"layout":{"autotypenumbers":"strict","colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"hovermode":"closest","hoverlabel":{"align":"left"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"bgcolor":"#E5ECF6","angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"ternary":{"bgcolor":"#E5ECF6","aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]]},"xaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"yaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"geo":{"bgcolor":"white","landcolor":"#E5ECF6","subunitcolor":"white","showland":true,"showlakes":true,"lakecolor":"white"},"title":{"x":0.05},"mapbox":{"style":"light"}}},"legend":{"tracegroupgap":0},"title":{"text":"Diverse Cancer subtype"}},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('4743743c-77b7-4b27-8fa4-2fe05a1be638');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                });            </script>        </div></div>
</div>
</section>
<section id="The-Sankey-diagram-delineates-the-relationships">
<h2>The Sankey diagram delineates the relationships<a class="headerlink" href="#The-Sankey-diagram-delineates-the-relationships" title="Link to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">aggregate_by_sum</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column_name</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">14</span><span class="p">):</span>

    <span class="c1"># Group by the specified column and calculate the sum for each group</span>
    <span class="n">group_counts</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">column_name</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>

    <span class="c1"># Select groups with a sum greater than the threshold</span>
    <span class="n">processed_data</span> <span class="o">=</span> <span class="n">group_counts</span><span class="p">[</span><span class="n">group_counts</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Filter the original DataFrame for rows belonging to the filtered groups</span>
    <span class="n">processed_data</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">processed_data</span><span class="p">)]</span>

    <span class="k">return</span> <span class="n">processed_data</span>

<span class="c1"># 使用函数</span>
<span class="n">result_df</span> <span class="o">=</span> <span class="n">aggregate_by_sum</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">result_df</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Automatic_MS_filename</th>
      <th>Batch</th>
      <th>Date</th>
      <th>Instrument</th>
      <th>Cell_line</th>
      <th>SIDM</th>
      <th>Tissue_type</th>
      <th>Cancer_type</th>
      <th>Cancer_subtype</th>
      <th>Project_Identifier</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>527</th>
      <td>181010_e0022_p02_2178_1_s_m04_1</td>
      <td>P02</td>
      <td>2018-10-10</td>
      <td>M04</td>
      <td>SW1710</td>
      <td>SIDM00420</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Bladder carcinoma</td>
      <td>SIDM00420;SW1710</td>
    </tr>
    <tr>
      <th>528</th>
      <td>181010_e0022_p02_2178_3_s_m04_1</td>
      <td>P02</td>
      <td>2018-10-10</td>
      <td>M04</td>
      <td>SW1710</td>
      <td>SIDM00420</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Bladder carcinoma</td>
      <td>SIDM00420;SW1710</td>
    </tr>
    <tr>
      <th>529</th>
      <td>181010_e0022_p02_2178_2_s_m04_1</td>
      <td>P02</td>
      <td>2018-10-10</td>
      <td>M04</td>
      <td>SW1710</td>
      <td>SIDM00420</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Bladder carcinoma</td>
      <td>SIDM00420;SW1710</td>
    </tr>
    <tr>
      <th>530</th>
      <td>181127_e0022_p02_2051_3_s_m04_1</td>
      <td>P02</td>
      <td>2018-11-27</td>
      <td>M04</td>
      <td>RT-112</td>
      <td>SIDM00402</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Bladder carcinoma</td>
      <td>SIDM00402;RT-112</td>
    </tr>
    <tr>
      <th>531</th>
      <td>181127_e0022_p02_2051_1_s_m04_1</td>
      <td>P02</td>
      <td>2018-11-27</td>
      <td>M04</td>
      <td>RT-112</td>
      <td>SIDM00402</td>
      <td>Bladder</td>
      <td>Bladder Carcinoma</td>
      <td>Bladder carcinoma</td>
      <td>SIDM00402;RT-112</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6590</th>
      <td>200131_b4-7-t3-1_00q3n_00rtc_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>COR-L95</td>
      <td>SIDM00521</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00521;COR-L95</td>
    </tr>
    <tr>
      <th>6591</th>
      <td>200131_b4-9-t3-1_00q3p_00rte_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H510A</td>
      <td>SIDM00927</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00927;NCI-H510A</td>
    </tr>
    <tr>
      <th>6592</th>
      <td>200131_b4-10-t3-1_00q3q_00rtf_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H2171</td>
      <td>SIDM00733</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00733;NCI-H2171</td>
    </tr>
    <tr>
      <th>6593</th>
      <td>200131_b4-13-t3-1_00q3t_00rti_m05_s_1</td>
      <td>P06</td>
      <td>2020-01-31</td>
      <td>M05</td>
      <td>NCI-H1836</td>
      <td>SIDM00770</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00770;NCI-H1836</td>
    </tr>
    <tr>
      <th>6594</th>
      <td>200201_b3-13-t3-2_00q3d_00rtm_m05_s_1</td>
      <td>P06</td>
      <td>2020-02-01</td>
      <td>M05</td>
      <td>IST-SL1</td>
      <td>SIDM00223</td>
      <td>Lung</td>
      <td>Small Cell Lung Carcinoma</td>
      <td>Small cell lung carcinoma</td>
      <td>SIDM00223;IST-SL1</td>
    </tr>
  </tbody>
</table>
<p>1188 rows × 10 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">generate_color_map</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">alpha_nodes</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">alpha_links</span><span class="o">=</span><span class="mf">0.55</span><span class="p">):</span>
    <span class="n">hues</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))]</span>
    <span class="n">colors_hsv</span> <span class="o">=</span> <span class="p">[(</span><span class="n">h</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">)</span> <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">hues</span><span class="p">]</span>

    <span class="n">colors_rgba_nodes</span> <span class="o">=</span> <span class="p">[</span><span class="n">colorsys</span><span class="o">.</span><span class="n">hsv_to_rgb</span><span class="p">(</span><span class="o">*</span><span class="n">hsv</span><span class="p">)</span> <span class="k">for</span> <span class="n">hsv</span> <span class="ow">in</span> <span class="n">colors_hsv</span><span class="p">]</span>
    <span class="n">colors_rgba_links</span> <span class="o">=</span> <span class="p">[(</span><span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">alpha_links</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">colors_rgba_nodes</span><span class="p">]</span>
    <span class="n">colors_rgba_nodes</span> <span class="o">=</span> <span class="p">[(</span><span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">alpha_nodes</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">colors_rgba_nodes</span><span class="p">]</span>

    <span class="n">colors_str_nodes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rgba(</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">g</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">b</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="n">a</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">colors_rgba_nodes</span><span class="p">]</span>
    <span class="n">colors_str_links</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rgba(</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">,</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">g</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">b</span><span class="o">*</span><span class="mi">255</span><span class="p">),</span> <span class="n">a</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">colors_rgba_links</span><span class="p">]</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">colors_str_nodes</span><span class="p">)),</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">colors_str_links</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">draw_sankey</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">all_labels</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())))</span>

    <span class="n">node_color_map</span><span class="p">,</span> <span class="n">link_color_map</span> <span class="o">=</span> <span class="n">generate_color_map</span><span class="p">(</span><span class="n">all_labels</span><span class="p">)</span>

    <span class="c1"># Aggregate data</span>
    <span class="n">links</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">,</span> <span class="s1">&#39;Cancer_type&#39;</span><span class="p">,</span> <span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">)</span>

    <span class="n">final_source</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">final_target</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">final_value</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">links</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">final_source</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">all_labels</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Tissue_type&#39;</span><span class="p">]))</span>
        <span class="n">final_target</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">all_labels</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">]))</span>
        <span class="n">final_value</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">])</span>

        <span class="n">final_source</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">all_labels</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">]))</span>
        <span class="n">final_target</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">all_labels</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Cancer_subtype&#39;</span><span class="p">]))</span>
        <span class="n">final_value</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">])</span>

    <span class="n">link_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">link_color_map</span><span class="p">[</span><span class="n">all_labels</span><span class="p">[</span><span class="n">s</span><span class="p">]]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">final_source</span><span class="p">]</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Sankey</span><span class="p">(</span>
        <span class="n">node</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">pad</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
                <span class="n">thickness</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
                <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
                <span class="n">label</span><span class="o">=</span><span class="n">all_labels</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">node_color_map</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">all_labels</span><span class="p">]</span>
        <span class="p">),</span>
        <span class="n">link</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">source</span><span class="o">=</span><span class="n">final_source</span><span class="p">,</span>
                <span class="n">target</span><span class="o">=</span><span class="n">final_target</span><span class="p">,</span>
                <span class="n">value</span><span class="o">=</span><span class="n">final_value</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="n">link_colors</span>
            <span class="p">)</span>
    <span class="p">))</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">title_text</span><span class="o">=</span><span class="s2">&quot;Sankey Diagram of Cancer Types and Subtypes&quot;</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">filename</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># 使用示例</span>
<span class="c1"># df = pd.DataFrame(data)</span>
<span class="n">draw_sankey</span><span class="p">(</span><span class="n">result_df</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sankey_diagram-3.pdf&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>                            <div id="b644e8e2-9bc7-425c-a0b0-53883319d284" class="plotly-graph-div" style="height:525px; width:100%;"></div>            <script type="text/javascript">                require(["plotly"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("b644e8e2-9bc7-425c-a0b0-53883319d284")) {                    Plotly.newPlot(                        "b644e8e2-9bc7-425c-a0b0-53883319d284",                        [{"link":{"color":["rgba(204,102,102,0.55)","rgba(204,119,102,0.55)","rgba(204,171,102,0.55)","rgba(102,204,142,0.55)","rgba(102,204,195,0.55)","rgba(102,195,204,0.55)","rgba(102,204,195,0.55)","rgba(102,195,204,0.55)","rgba(102,160,204,0.55)","rgba(183,204,102,0.55)","rgba(102,160,204,0.55)","rgba(183,204,102,0.55)","rgba(102,160,204,0.55)","rgba(183,204,102,0.55)","rgba(102,160,204,0.55)","rgba(183,204,102,0.55)","rgba(102,142,204,0.55)","rgba(102,204,160,0.55)","rgba(102,107,204,0.55)","rgba(113,102,204,0.55)","rgba(102,107,204,0.55)","rgba(204,102,171,0.55)","rgba(131,102,204,0.55)","rgba(148,102,204,0.55)","rgba(131,102,204,0.55)","rgba(148,102,204,0.55)","rgba(131,102,204,0.55)","rgba(148,102,204,0.55)","rgba(204,102,119,0.55)","rgba(148,204,102,0.55)","rgba(204,102,119,0.55)","rgba(148,204,102,0.55)","rgba(204,102,119,0.55)","rgba(148,204,102,0.55)","rgba(204,102,119,0.55)","rgba(148,204,102,0.55)"],"source":[0,1,4,14,17,18,17,18,20,7,20,7,20,7,20,7,21,15,23,24,23,31,25,26,25,26,25,26,34,9,34,9,34,9,34,9],"target":[1,2,14,22,18,5,18,19,7,3,7,6,7,8,7,30,15,16,24,33,31,32,26,27,26,28,26,29,9,10,9,11,9,12,9,13],"value":[89,89,72,72,64,64,107,107,46,46,147,147,43,43,35,35,72,72,86,86,187,187,44,44,13,13,52,52,75,75,12,12,29,29,15,15]},"node":{"color":["rgba(204,102,102,1.0)","rgba(204,119,102,1.0)","rgba(204,136,102,1.0)","rgba(204,154,102,1.0)","rgba(204,171,102,1.0)","rgba(204,189,102,1.0)","rgba(201,204,102,1.0)","rgba(183,204,102,1.0)","rgba(166,204,102,1.0)","rgba(148,204,102,1.0)","rgba(131,204,102,1.0)","rgba(113,204,102,1.0)","rgba(102,204,107,1.0)","rgba(102,204,125,1.0)","rgba(102,204,142,1.0)","rgba(102,204,160,1.0)","rgba(102,204,177,1.0)","rgba(102,204,195,1.0)","rgba(102,195,204,1.0)","rgba(102,177,204,1.0)","rgba(102,160,204,1.0)","rgba(102,142,204,1.0)","rgba(102,125,204,1.0)","rgba(102,107,204,1.0)","rgba(113,102,204,1.0)","rgba(131,102,204,1.0)","rgba(148,102,204,1.0)","rgba(166,102,204,1.0)","rgba(183,102,204,1.0)","rgba(201,102,204,1.0)","rgba(204,102,189,1.0)","rgba(204,102,171,1.0)","rgba(204,102,154,1.0)","rgba(204,102,136,1.0)","rgba(204,102,119,1.0)"],"label":["Bladder","Bladder Carcinoma","Bladder carcinoma","Cecum adenocarcinoma","Central Nervous System","Clear cell renal cell carcinoma","Colon adenocarcinoma","Colorectal Carcinoma","Colorectal carcinoma","Gastric Carcinoma","Gastric adenocarcinoma","Gastric carcinoma","Gastric signet ring cell adenocarcinoma","Gastric tubular adenocarcinoma","Glioma","Hepatocellular Carcinoma","Hepatocellular carcinoma","Kidney","Kidney Carcinoma","Kidney carcinoma","Large Intestine","Liver","Low grade glioma","Lung","Non-Small Cell Lung Carcinoma","Pancreas","Pancreatic Carcinoma","Pancreatic adenocarcinoma","Pancreatic carcinoma","Pancreatic ductal adenocarcinoma","Rectal adenocarcinoma","Small Cell Lung Carcinoma","Small cell lung carcinoma","Squamous cell lung carcinoma","Stomach"],"line":{"color":"black","width":0.5},"pad":15,"thickness":20},"type":"sankey"}],                        {"template":{"data":{"histogram2dcontour":[{"type":"histogram2dcontour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"choropleth":[{"type":"choropleth","colorbar":{"outlinewidth":0,"ticks":""}}],"histogram2d":[{"type":"histogram2d","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmap":[{"type":"heatmap","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmapgl":[{"type":"heatmapgl","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"contourcarpet":[{"type":"contourcarpet","colorbar":{"outlinewidth":0,"ticks":""}}],"contour":[{"type":"contour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"surface":[{"type":"surface","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"mesh3d":[{"type":"mesh3d","colorbar":{"outlinewidth":0,"ticks":""}}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"parcoords":[{"type":"parcoords","line":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolargl":[{"type":"scatterpolargl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"scattergeo":[{"type":"scattergeo","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolar":[{"type":"scatterpolar","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"scattergl":[{"type":"scattergl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatter3d":[{"type":"scatter3d","line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermapbox":[{"type":"scattermapbox","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterternary":[{"type":"scatterternary","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattercarpet":[{"type":"scattercarpet","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"pie":[{"automargin":true,"type":"pie"}]},"layout":{"autotypenumbers":"strict","colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"hovermode":"closest","hoverlabel":{"align":"left"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"bgcolor":"#E5ECF6","angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"ternary":{"bgcolor":"#E5ECF6","aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]]},"xaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"yaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"geo":{"bgcolor":"white","landcolor":"#E5ECF6","subunitcolor":"white","showland":true,"showlakes":true,"lakecolor":"white"},"title":{"x":0.05},"mapbox":{"style":"light"}}},"title":{"text":"Sankey Diagram of Cancer Types and Subtypes"},"font":{"size":10}},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('b644e8e2-9bc7-425c-a0b0-53883319d284');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                });            </script>        </div></div>
</div>
</section>
<section id="The-reproducibility-and-reliability-of-the-STAVER-processed-data">
<h2>The reproducibility and reliability of the STAVER-processed data<a class="headerlink" href="#The-reproducibility-and-reliability-of-the-STAVER-processed-data" title="Link to this heading">¶</a></h2>
<section id="HEK293T-Spearman-correlation-analysis">
<h3>HEK293T Spearman correlation analysis<a class="headerlink" href="#HEK293T-Spearman-correlation-analysis" title="Link to this heading">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load HEK293T data</span>
<span class="n">HEK293T_rawdata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/HEK_293T/HEK-QCS_matrix_rawdata.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">HEK293T_rawdata</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">thresh</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">HEK293T_STAVER</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/HEK_293T/HEK-QCS_matrix_STAVER.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">HEK293T_STAVER</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">thresh</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">correlation_heatmap</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;pearson&#39;</span><span class="p">,</span> <span class="n">log_transformed</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">outpath</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Correaltion heatmap and correlation matrix</span>
<span class="sd">    Args:</span>
<span class="sd">    --------------</span>
<span class="sd">    data -&gt; Dataframe: dataframe of raw data</span>
<span class="sd">    method -&gt; str: pearson or spearman</span>
<span class="sd">    log_transformed -&gt; bool: whether to use log10 transformed data</span>

<span class="sd">    Return:</span>
<span class="sd">    -----------</span>
<span class="sd">    Dataframe: Correaltion of experiments data</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">log_transformed</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="n">corr</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span> <span class="o">=</span> <span class="n">method</span><span class="p">)</span>
    <span class="c1">## Platelet selection refernce: https://learnku.com/articles/39890</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="c1"># optional cmap: RdYlGn; RdYlGn_r; YlGn; rocket_r; YlGnBu; YlGnBu_r; YlOrBr; YlOrBr_r; YlOrRd; YlOrRd_r; RdBu_r</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">corr</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Reds&quot;</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">col_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">row_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">.005</span><span class="p">,</span><span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;.2f&quot;</span><span class="p">)</span>
    <span class="c1"># sn.heatmap(corr, annot=True, cmap=&#39;vlag&#39;)  ## optional cmap: RdYlGn; RdYlGn_r; YlGn; rocket_r; YlGnBu; YlGnBu_r; YlOrBr; YlOrBr_r; YlOrRd; YlOrRd_r;</span>
    <span class="k">if</span> <span class="n">outpath</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">_corr_heatmap.pdf&quot;</span><span class="p">)</span>
        <span class="n">corr</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outpath</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">_corr_matrix.csv&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="c1"># return corr</span>
</pre></div>
</div>
</div>
</section>
<section id="The-correlation-heatmap-of-raw-data">
<h3>The correlation heatmap of raw data<a class="headerlink" href="#The-correlation-heatmap-of-raw-data" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outpath</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;~/STAVER-revised/figs/&#39;</span>
<span class="n">correlation_heatmap</span><span class="p">(</span><span class="n">HEK293T_rawdata</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;spearman&#39;</span><span class="p">,</span> <span class="n">log_transformed</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">outpath</span> <span class="o">=</span> <span class="n">outpath</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;HEK-QCS_matrix_rawdata_corr_heatmap&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Figure size 1000x1000 with 0 Axes&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_21_1.png" src="_images/PanCancer_949_celline_analysis_21_1.png" />
</div>
</div>
</section>
<section id="The-correlation-heatmap-of-STAVER-processed-data">
<h3>The correlation heatmap of STAVER-processed data<a class="headerlink" href="#The-correlation-heatmap-of-STAVER-processed-data" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outpath</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;~/STAVER-revised/figs/&#39;</span>
<span class="n">correlation_heatmap</span><span class="p">(</span><span class="n">HEK293T_STAVER</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;spearman&#39;</span><span class="p">,</span> <span class="n">log_transformed</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">outpath</span> <span class="o">=</span> <span class="n">outpath</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;HEK-QCS_matrix_STAVER_corr_heatmap&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Figure size 1000x1000 with 0 Axes&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_23_1.png" src="_images/PanCancer_949_celline_analysis_23_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[93]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">extract_lower_triangular</span><span class="p">(</span><span class="n">df_corr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    提取相关性矩阵对角线以下的一半相关性值。</span>

<span class="sd">    参数:</span>
<span class="sd">    ----</span>
<span class="sd">    df_corr : pandas.DataFrame</span>
<span class="sd">        相关性矩阵的 DataFrame。</span>

<span class="sd">    返回:</span>
<span class="sd">    -----</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        包含对角线以下一半相关性值的 DataFrame。</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">df_corr</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;df_corr 必须是 pandas.DataFrame 类型。&quot;</span><span class="p">)</span>

    <span class="c1"># 获取对角线以下的索引</span>
    <span class="n">tril_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tril_indices</span><span class="p">(</span><span class="n">df_corr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">k</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># 提取对角线以下的一半相关性值</span>
    <span class="n">lower_triangular_values</span> <span class="o">=</span> <span class="n">df_corr</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()[</span><span class="n">tril_indices</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s1">&#39;Row&#39;</span><span class="p">:</span> <span class="n">df_corr</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">tril_indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span>
            <span class="s1">&#39;Column&#39;</span><span class="p">:</span> <span class="n">df_corr</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">tril_indices</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span>
            <span class="s1">&#39;Correlation&#39;</span><span class="p">:</span> <span class="n">lower_triangular_values</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">get_IQR</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="n">Q1</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">)</span>
    <span class="n">Q3</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>
    <span class="n">IQR</span> <span class="o">=</span> <span class="n">Q3</span> <span class="o">-</span> <span class="n">Q1</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The IQR1 is: </span><span class="si">{</span><span class="n">Q1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The IQR3 is: </span><span class="si">{</span><span class="n">Q3</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The mean is: </span><span class="si">{</span><span class="n">mean</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">median</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The median is: </span><span class="si">{</span><span class="n">median</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="The-IQR-of-HEK293T-spearman-correlation-matrix-in-rawdata">
<h3>The IQR of HEK293T spearman correlation matrix in rawdata<a class="headerlink" href="#The-IQR-of-HEK293T-spearman-correlation-matrix-in-rawdata" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[105]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">=</span> <span class="n">extract_lower_triangular</span><span class="p">(</span><span class="n">HEK293T_rawdata</span><span class="o">.</span><span class="n">corr</span><span class="p">())</span>
<span class="n">get_IQR</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;Correlation&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The IQR1 is: 0.8846399725
The IQR3 is: 0.907049235
The mean is: 0.893689345854834
The median is: 0.897238564
</pre></div></div>
</div>
</section>
<section id="The-IQR-of-HEK293T-spearman-correlation-matrix-in-SATVER-processed-data">
<h3>The IQR of HEK293T spearman correlation matrix in SATVER-processed data<a class="headerlink" href="#The-IQR-of-HEK293T-spearman-correlation-matrix-in-SATVER-processed-data" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[106]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">=</span> <span class="n">extract_lower_triangular</span><span class="p">(</span><span class="n">HEK293T_STAVER</span><span class="o">.</span><span class="n">corr</span><span class="p">())</span>
<span class="n">get_IQR</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;Correlation&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The IQR1 is: 0.9321159908757065
The IQR3 is: 0.9821799248475355
The mean is: 0.9523831119684066
The median is: 0.9657469442578246
</pre></div></div>
</div>
</section>
<section id="The-identification-protein-numbers-of-the-raw-data-and-STAVER-data">
<h3>The identification protein numbers of the raw data and STAVER data<a class="headerlink" href="#The-identification-protein-numbers-of-the-raw-data-and-STAVER-data" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protein_num</span><span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_clipboard</span><span class="p">()</span>
<span class="n">protein_num</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Experiment</th>
      <th>ldentification of Protein numbers</th>
      <th>Type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>190124_HEK-QCS_000F2_008AR_M04_S_1</td>
      <td>5683</td>
      <td>Raw data</td>
    </tr>
    <tr>
      <th>1</th>
      <td>190124_HEK-QCS_000F2_008IQ_M06_S_1</td>
      <td>5923</td>
      <td>Raw data</td>
    </tr>
    <tr>
      <th>2</th>
      <td>190125_HEK-QCS_000F2_008JO_M04_S_1</td>
      <td>5617</td>
      <td>Raw data</td>
    </tr>
    <tr>
      <th>3</th>
      <td>190128_HEK-QCS_000F2_008MR_M06_S_1</td>
      <td>5873</td>
      <td>Raw data</td>
    </tr>
    <tr>
      <th>4</th>
      <td>190129_HEK-QCS_000F2_008NM_M04_S_1</td>
      <td>5465</td>
      <td>Raw data</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>162</th>
      <td>190102_HEK-QCS_000F2_007K6_M02_S_1</td>
      <td>5150</td>
      <td>STAVER</td>
    </tr>
    <tr>
      <th>163</th>
      <td>190401_HEK-QCS_000F2_00A13_M06_S_1</td>
      <td>5817</td>
      <td>STAVER</td>
    </tr>
    <tr>
      <th>164</th>
      <td>190507_HEK-QCS_000F2_00AQK_M03_S_1</td>
      <td>5961</td>
      <td>STAVER</td>
    </tr>
    <tr>
      <th>165</th>
      <td>190522_HEK-QCS_000F2_00BEO_M04_S_1</td>
      <td>5283</td>
      <td>STAVER</td>
    </tr>
    <tr>
      <th>166</th>
      <td>181213_HEK-QCS_000F2_006UR_M03_S_1</td>
      <td>5519</td>
      <td>STAVER</td>
    </tr>
  </tbody>
</table>
<p>167 rows × 3 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_protein_numbers</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span> <span class="n">y_lim</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="mf">1.3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a combined box and strip plot for protein number visualization.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pd.DataFrame): DataFrame containing the data to be plotted.</span>
<span class="sd">        x_col (str): Column name in `data` to be plotted on the x-axis.</span>
<span class="sd">        y_col (str): Column name in `data` to be plotted on the y-axis.</span>
<span class="sd">        title (str): Title of the plot.</span>
<span class="sd">        y_lim (tuple, optional): Tuple specifying the limits for the y-axis (e.g., (0, 7000)).</span>
<span class="sd">        height (float, optional): Height (in inches) of each facet. Defaults to 6.</span>
<span class="sd">        aspect (float, optional): Aspect ratio of each facet, so that aspect * height gives the width of each facet in inches. Defaults to 1.3.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; plot_protein_numbers(protein_num, &#39;Type&#39;, &#39;Identification of Protein numbers&#39;,</span>
<span class="sd">                                &#39;Protein numbers of HEK293T QC samples&#39;, y_lim=(0, 7000), height=5, aspect=1)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">custom_params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;axes.spines.right&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;axes.spines.top&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set_theme</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;ticks&quot;</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="n">custom_params</span><span class="p">)</span>

    <span class="c1"># Create a box plot</span>
    <span class="n">box_plot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">catplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x_col</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y_col</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">x_col</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;box&quot;</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="n">aspect</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Overlay with a strip plot</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x_col</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y_col</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">x_col</span><span class="p">,</span> <span class="n">jitter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dodge</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set2&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Set additional plot attributes</span>
    <span class="n">box_plot</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>  <span class="c1"># Set the title for the figure</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">y_lim</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">y_lim</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># The protein numbers of the raw data and STAVER data</span>
<span class="n">plot_protein_numbers</span><span class="p">(</span><span class="n">protein_num</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span> <span class="s1">&#39;ldentification of Protein numbers&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;Identification of the raw and STAVER data&#39;</span><span class="p">,</span> <span class="n">y_lim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">7000</span><span class="p">),</span> <span class="n">height</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_31_0.png" src="_images/PanCancer_949_celline_analysis_31_0.png" />
</div>
</div>
</section>
<section id="The-Coefficient-of-Variation-(CVs)-of-the-raw-data-and-STAVER-data">
<h3>The Coefficient of Variation (CVs) of the raw data and STAVER data<a class="headerlink" href="#The-Coefficient-of-Variation-(CVs)-of-the-raw-data-and-STAVER-data" title="Link to this heading">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[130]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># custom_params = {&quot;axes.spines.right&quot;: False, &quot;axes.spines.top&quot;: False}</span>
<span class="c1"># sns.set_theme(style=&quot;ticks&quot;, rc=custom_params)</span>

<span class="k">def</span> <span class="nf">plot_molecular_variance</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column_name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the density curves of the original and STAVER processed data for comparison.</span>

<span class="sd">    Args:</span>
<span class="sd">        df: A pandas dataframe containing the data.</span>
<span class="sd">        column_name: A string representing the column name of the data to be plotted.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

    <span class="c1"># Density plot of the original data</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Raw data&#39;</span><span class="p">][</span><span class="s1">&#39;Coefficient of Variation [%]&#39;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Original Density&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>

    <span class="c1"># Density plot of the STAVER-processed data</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;STAVER&#39;</span><span class="p">][</span><span class="s1">&#39;Coefficient of Variation [%]&#39;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;STAVER Density&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Coefficient of Variation density curve&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Coefficient of Variation [%]&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[131]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protein_cv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/Pan-cancer-cell-lines/HEK293T_Protein_CV_Compare.csv&quot;</span><span class="p">)</span>
<span class="n">plot_molecular_variance</span><span class="p">(</span><span class="n">protein_cv</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_34_0.png" src="_images/PanCancer_949_celline_analysis_34_0.png" />
</div>
</div>
</section>
</section>
<section id="The-advantages-of-the-STAVER-algorithm-to-uncover-inherent-biological-differences">
<h2>The advantages of the STAVER algorithm to uncover inherent biological differences<a class="headerlink" href="#The-advantages-of-the-STAVER-algorithm-to-uncover-inherent-biological-differences" title="Link to this heading">¶</a></h2>
<p>Distinct tumors often exhibit diverse molecular characteristics <strong>(Cell, 2014, PMID: 25109877; Cell, 2023, PMID: 37582357)</strong>, with even different histological subtypes of the same tumor demonstrating significant molecular heterogeneity <strong>(Science, 2014, PMID: 25301631; Cancer Cell, 2023, PMID: 36563681)</strong>, contributing to challenges in cancer treatment. Thus, accurately deciphering the inherent molecular heterogeneity of various cancer types, especially in high-dimensional data such as
proteomics, is vital for precision treatments and ultimately improved patient outcomes.</p>
<p>To demonstrate the superior advantages of the STAVER algorithm to uncover inherent biological differences, we comprehensively compared the original data and STAVER-processed data of the 1,242 cancer cell line samples from diverse cancer types.</p>
<section id="The-UMAP-analysis-of-diverse-1242-cancer-cell-line-samples">
<h3>The UMAP analysis of diverse 1242 cancer cell line samples<a class="headerlink" href="#The-UMAP-analysis-of-diverse-1242-cancer-cell-line-samples" title="Link to this heading">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[98]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">visualize_cancer_subtypes</span><span class="p">(</span><span class="n">proteomics_data_path</span><span class="p">,</span>
                              <span class="n">subtypes_file_path</span><span class="p">,</span>
                              <span class="n">color_params</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Batch&#39;</span><span class="p">,</span> <span class="s1">&#39;Tissue_type&#39;</span><span class="p">,</span> <span class="s1">&#39;Cancer_type&#39;</span><span class="p">],</span>
                              <span class="n">figsize_per_plot</span><span class="o">=</span><span class="p">(</span><span class="mf">4.5</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                              <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">,</span>
                              <span class="n">outpath</span><span class="o">=</span><span class="s1">&#39;./&#39;</span><span class="p">,</span>
                              <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                              <span class="n">add_outline</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                              <span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize the given proteomics data and cancer subtypes using t-SNE and UMAP.</span>

<span class="sd">    Args:</span>
<span class="sd">        proteomics_data_path (str): Path to the proteomics data (samples as rows, proteins as columns).</span>
<span class="sd">        subtypes_file_path (str): Path to the file with samples and their corresponding cancer subtypes.</span>
<span class="sd">        color_params (list): List of parameters to color by in the plots.</span>
<span class="sd">        figsize_per_plot (tuple): Size of each individual plot.</span>
<span class="sd">        outpath (str): Directory to save the plots.</span>
<span class="sd">        filename (str): Prefix of the saved files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None. Generates visualization images for t-SNE and UMAP.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">proteomics_data_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Load subtype information</span>
    <span class="n">subtypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">subtypes_file_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Check for missing subtype information</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">sample</span> <span class="ow">in</span> <span class="n">subtypes</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Some samples lack subtype information!&quot;</span><span class="p">)</span>

    <span class="c1"># Convert data to AnnData format</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Add subtype information to AnnData</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">subtypes</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span>

    <span class="c1"># Data normalization (if required)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

    <span class="c1"># Compute t-SNE and UMAP</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

    <span class="c1"># Set filenames if not provided</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">filename</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">proteomics_data_path</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># # Visualize t-SNE</span>
    <span class="c1"># fig, axes = plt.subplots(figsize=(len(color_params)*figsize_per_plot[0], figsize_per_plot[1]), nrows=1, ncols=len(color_params))</span>
    <span class="c1"># for i, param in enumerate(color_params):</span>
    <span class="c1">#     sc.pl.tsne(adata, color=param, legend_loc=legend_loc, title=f&quot;t-SNE colored by {param}&quot;, ax=axes[i], show=show_plot)</span>
    <span class="c1"># fig.tight_layout()</span>
    <span class="c1"># fig.savefig(os.path.join(outpath, f&quot;{filename}_tsne.pdf&quot;))</span>
    <span class="c1"># plt.close(fig)</span>

    <span class="c1"># # Visualize UMAP</span>
    <span class="c1"># fig, axes = plt.subplots(figsize=(len(color_params)*figsize_per_plot[0], figsize_per_plot[1]), nrows=1, ncols=len(color_params))</span>
    <span class="c1"># for i, param in enumerate(color_params):</span>
    <span class="c1">#     sc.pl.umap(adata, color=param, legend_loc=legend_loc, title=f&quot;UMAP colored by {param}&quot;, ax=axes[i], show=show_plot)</span>
    <span class="c1"># fig.tight_layout()</span>
    <span class="c1"># fig.savefig(os.path.join(outpath, f&quot;{filename}_umap.pdf&quot;))</span>
    <span class="c1"># plt.close(fig)</span>

    <span class="c1"># Determine the number of plots (number of color_params x 2 for both t-SNE and UMAP)</span>
    <span class="n">nrows</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">ncols</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">color_params</span><span class="p">)</span>

    <span class="c1"># Create a figure with the necessary number of subplots</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="n">ncols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">ncols</span> <span class="o">*</span> <span class="n">figsize_per_plot</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">nrows</span> <span class="o">*</span> <span class="n">figsize_per_plot</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

    <span class="c1"># Visualize t-SNE on the first row</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">color_params</span><span class="p">):</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">param</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;t-SNE colored by </span><span class="si">{</span><span class="n">param</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">add_outline</span><span class="o">=</span><span class="n">add_outline</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">i</span><span class="p">],</span> <span class="n">show</span><span class="o">=</span><span class="n">show_plot</span><span class="p">)</span>

    <span class="c1"># Visualize UMAP on the second row</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">color_params</span><span class="p">):</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">param</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;UMAP colored by </span><span class="si">{</span><span class="n">param</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">add_outline</span><span class="o">=</span><span class="n">add_outline</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">i</span><span class="p">],</span> <span class="n">show</span><span class="o">=</span><span class="n">show_plot</span><span class="p">)</span>

    <span class="c1"># Save the combined visualization</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">_combined.pdf&quot;</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>
</pre></div>
</div>
</div>
</section>
<section id="The-UMAP-analysis-of-the-Rawdata">
<h3>The UMAP analysis of the Rawdata<a class="headerlink" href="#The-UMAP-analysis-of-the-Rawdata" title="Link to this heading">¶</a></h3>
<p>By employing UMAP analysis, the results indicated that the STAVER algorithm did not introduce batch effects, corroborating the original data findings <strong>(Figure RL2H-2I and Figure RL5H-5I)</strong>. In the original data, the UMAP analysis showed that the 1,242 cancer cell line samples were diffusely distributed across various tissue sources and cancer types and did not show a clear separation. This lack of distinction was particularly evidenced among bladder, colorectal, pancreatic, gastric, and
hepatocellular cancer cell lines, which intermixed with each other <strong>(Figure RL2H and Figure RL5H)</strong>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[100]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outpath</span> <span class="o">=</span> <span class="s1">&#39;~/DIA-STAVER/STAVER-nc-revised/revised-sript/figs-1/&#39;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outpath</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outpath</span><span class="p">)</span>
<span class="n">adata_rawdata</span> <span class="o">=</span> <span class="n">visualize_cancer_subtypes</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/PCA/rawdata_data.csv&quot;</span><span class="p">,</span> <span class="s2">&quot;~/STAVER-revised/PCA/metadata_anatation_scanpy.csv&quot;</span><span class="p">,</span> <span class="n">outpath</span><span class="o">=</span><span class="n">outpath</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;Raw_data_2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: You’re trying to run this on 11880 dimensions of `.X`, if you really want this, set `use_rep=&#39;X&#39;`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
OMP: Info #271: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_39_2.png" src="_images/PanCancer_949_celline_analysis_39_2.png" />
</div>
</div>
</section>
<section id="The-UMAP-analysis-of-STAVER-processed-data">
<h3>The UMAP analysis of STAVER-processed data<a class="headerlink" href="#The-UMAP-analysis-of-STAVER-processed-data" title="Link to this heading">¶</a></h3>
<p>When analyzing STAVER-processed data, UMAP visualization clearly separated the 1,242 cancer cell line samples by tissue sources and cancer types <strong>(Figure RL2I and Figure RL5I)</strong>. Notably, the STAVER-processed data revealed that cancer cell lines from the digestive tract, such as intestinal, gastric, and pancreatic, exhibited more molecular similarities than their nondigestive tract counterparts (lung, glioma, and kidney cancer cell lines). These findings highlighted the potential superiority of
the STAVER algorithm in elucidating the intrinsic biological differences among diverse tumor cell lines.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[102]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outpath</span> <span class="o">=</span> <span class="s1">&#39;~/DIA-STAVER/STAVER-nc-revised/revised-sript/figs-1/&#39;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outpath</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outpath</span><span class="p">)</span>

<span class="n">adata_STAVER</span> <span class="o">=</span> <span class="n">visualize_cancer_subtypes</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/PCA/STAVER_data.csv&quot;</span><span class="p">,</span> <span class="s2">&quot;~/STAVER-revised/PCA/metadata_anatation_scanpy.csv&quot;</span><span class="p">,</span> <span class="n">outpath</span><span class="o">=</span><span class="n">outpath</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;STAVER_data_2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: You’re trying to run this on 8202 dimensions of `.X`, if you really want this, set `use_rep=&#39;X&#39;`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_41_1.png" src="_images/PanCancer_949_celline_analysis_41_1.png" />
</div>
</div>
</section>
<section id="Cancer-specific-proteins">
<h3>Cancer specific proteins<a class="headerlink" href="#Cancer-specific-proteins" title="Link to this heading">¶</a></h3>
<p>To rigorously establish the advantages of the STAVER algorithm in identifying tumor cell-specific proteins, we compared the distribution patterns of cell-specific protein expression profiles between the original and STAVER-processed data. We incorporated the <a class="reference external" href="https://www.proteinatlas.org">Human Protein Atlas (HPA)</a> molecular annotations of tumor cell specificity and their expression profiles across various cancer cell lines for integrated analysis.</p>
<p>The results demonstrated that the STAVER-processed data more accurately reflected cancer type-specific expression patterns than the original proteomic data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HPA_dataset</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-nc-revised/resources/proteinatlas_56a92855.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">HPA_dataset</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene</th>
      <th>Gene synonym</th>
      <th>Ensembl</th>
      <th>Gene description</th>
      <th>Uniprot</th>
      <th>Chromosome</th>
      <th>Position</th>
      <th>Protein class</th>
      <th>Biological process</th>
      <th>Molecular function</th>
      <th>...</th>
      <th>Pathology prognostics - Lung cancer</th>
      <th>Pathology prognostics - Melanoma</th>
      <th>Pathology prognostics - Ovarian cancer</th>
      <th>Pathology prognostics - Pancreatic cancer</th>
      <th>Pathology prognostics - Prostate cancer</th>
      <th>Pathology prognostics - Renal cancer</th>
      <th>Pathology prognostics - Stomach cancer</th>
      <th>Pathology prognostics - Testis cancer</th>
      <th>Pathology prognostics - Thyroid cancer</th>
      <th>Pathology prognostics - Urothelial cancer</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A1BG</td>
      <td>NaN</td>
      <td>ENSG00000121410</td>
      <td>Alpha-1-B glycoprotein</td>
      <td>P04217</td>
      <td>19</td>
      <td>58345178-58353492</td>
      <td>Plasma proteins, Predicted intracellular prote...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>unprognostic (1.09e-1)</td>
      <td>unprognostic (2.59e-1)</td>
      <td>unprognostic (2.10e-1)</td>
      <td>unprognostic (1.47e-2)</td>
      <td>unprognostic (1.37e-2)</td>
      <td>unprognostic (4.19e-5)</td>
      <td>unprognostic (2.37e-2)</td>
      <td>unprognostic (1.94e-1)</td>
      <td>unprognostic (1.72e-1)</td>
      <td>unprognostic (6.72e-2)</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A1CF</td>
      <td>ACF, ACF64, ACF65, APOBEC1CF, ASP</td>
      <td>ENSG00000148584</td>
      <td>APOBEC1 complementation factor</td>
      <td>Q9NQ94</td>
      <td>10</td>
      <td>50799409-50885675</td>
      <td>Predicted intracellular proteins</td>
      <td>mRNA processing</td>
      <td>RNA-binding</td>
      <td>...</td>
      <td>unprognostic (7.38e-3)</td>
      <td>NaN</td>
      <td>unprognostic (1.30e-2)</td>
      <td>unprognostic (2.46e-2)</td>
      <td>unprognostic (1.20e-1)</td>
      <td>unprognostic (1.90e-3)</td>
      <td>unprognostic (1.97e-2)</td>
      <td>unprognostic (2.77e-1)</td>
      <td>unprognostic (2.19e-2)</td>
      <td>unprognostic (8.50e-4)</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A2M</td>
      <td>CPAMD5, FWP007, S863-7</td>
      <td>ENSG00000175899</td>
      <td>Alpha-2-macroglobulin</td>
      <td>P01023</td>
      <td>12</td>
      <td>9067664-9116229</td>
      <td>Cancer-related genes, Candidate cardiovascular...</td>
      <td>NaN</td>
      <td>Protease inhibitor, Serine protease inhibitor</td>
      <td>...</td>
      <td>unprognostic (3.65e-2)</td>
      <td>unprognostic (2.38e-1)</td>
      <td>unprognostic (7.19e-2)</td>
      <td>unprognostic (4.71e-2)</td>
      <td>unprognostic (2.06e-2)</td>
      <td>unprognostic (1.28e-2)</td>
      <td>unprognostic (8.04e-3)</td>
      <td>unprognostic (2.32e-2)</td>
      <td>unprognostic (8.58e-2)</td>
      <td>unprognostic (9.03e-3)</td>
    </tr>
    <tr>
      <th>3</th>
      <td>A2ML1</td>
      <td>CPAMD9, FLJ25179, p170</td>
      <td>ENSG00000166535</td>
      <td>Alpha-2-macroglobulin like 1</td>
      <td>A8K2U0</td>
      <td>12</td>
      <td>8822621-8887001</td>
      <td>Disease related genes, Predicted intracellular...</td>
      <td>NaN</td>
      <td>Protease inhibitor, Serine protease inhibitor</td>
      <td>...</td>
      <td>unprognostic (7.58e-3)</td>
      <td>unprognostic (2.63e-1)</td>
      <td>unprognostic (1.57e-1)</td>
      <td>unprognostic (1.15e-3)</td>
      <td>unprognostic (2.03e-1)</td>
      <td>unprognostic (1.06e-9)</td>
      <td>unprognostic (2.28e-1)</td>
      <td>unprognostic (3.07e-1)</td>
      <td>unprognostic (5.88e-2)</td>
      <td>unprognostic (2.42e-2)</td>
    </tr>
    <tr>
      <th>4</th>
      <td>A3GALT2</td>
      <td>A3GALT2P, IGB3S, IGBS3S</td>
      <td>ENSG00000184389</td>
      <td>Alpha 1,3-galactosyltransferase 2</td>
      <td>U3KPV4</td>
      <td>1</td>
      <td>33306766-33321098</td>
      <td>Enzymes, Predicted membrane proteins</td>
      <td>Lipid metabolism</td>
      <td>Glycosyltransferase, Transferase</td>
      <td>...</td>
      <td>unprognostic (4.96e-2)</td>
      <td>unprognostic (6.83e-2)</td>
      <td>unprognostic (5.81e-2)</td>
      <td>unprognostic (1.23e-1)</td>
      <td>unprognostic (1.89e-1)</td>
      <td>unprognostic (4.90e-8)</td>
      <td>unprognostic (1.17e-1)</td>
      <td>NaN</td>
      <td>unprognostic (1.12e-2)</td>
      <td>unprognostic (7.87e-2)</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>20157</th>
      <td>ZYG11A</td>
      <td>ZYG11</td>
      <td>ENSG00000203995</td>
      <td>Zyg-11 family member A, cell cycle regulator</td>
      <td>Q6WRX3</td>
      <td>1</td>
      <td>52842511-52894998</td>
      <td>Predicted intracellular proteins</td>
      <td>Ubl conjugation pathway</td>
      <td>NaN</td>
      <td>...</td>
      <td>unprognostic (2.34e-1)</td>
      <td>unprognostic (4.56e-2)</td>
      <td>unprognostic (2.06e-2)</td>
      <td>unprognostic (4.01e-2)</td>
      <td>unprognostic (1.01e-1)</td>
      <td>unprognostic (6.15e-3)</td>
      <td>unprognostic (2.95e-1)</td>
      <td>unprognostic (1.21e-1)</td>
      <td>unprognostic (3.07e-1)</td>
      <td>unprognostic (1.02e-1)</td>
    </tr>
    <tr>
      <th>20158</th>
      <td>ZYG11B</td>
      <td>FLJ13456, ZYG11</td>
      <td>ENSG00000162378</td>
      <td>Zyg-11 family member B, cell cycle regulator</td>
      <td>Q9C0D3</td>
      <td>1</td>
      <td>52726453-52827336</td>
      <td>Predicted intracellular proteins</td>
      <td>Ubl conjugation pathway</td>
      <td>NaN</td>
      <td>...</td>
      <td>unprognostic (1.85e-1)</td>
      <td>unprognostic (4.84e-3)</td>
      <td>unprognostic (5.06e-2)</td>
      <td>unprognostic (2.76e-1)</td>
      <td>unprognostic (6.08e-2)</td>
      <td>prognostic favorable (9.80e-7)</td>
      <td>unprognostic (2.22e-1)</td>
      <td>unprognostic (3.37e-1)</td>
      <td>unprognostic (1.13e-1)</td>
      <td>unprognostic (9.57e-2)</td>
    </tr>
    <tr>
      <th>20159</th>
      <td>ZYX</td>
      <td>NaN</td>
      <td>ENSG00000159840</td>
      <td>Zyxin</td>
      <td>Q15942</td>
      <td>7</td>
      <td>143381295-143391111</td>
      <td>Plasma proteins, Predicted intracellular proteins</td>
      <td>Cell adhesion, Host-virus interaction</td>
      <td>NaN</td>
      <td>...</td>
      <td>unprognostic (1.66e-3)</td>
      <td>unprognostic (2.60e-1)</td>
      <td>unprognostic (4.22e-1)</td>
      <td>unprognostic (1.98e-1)</td>
      <td>unprognostic (2.43e-1)</td>
      <td>prognostic unfavorable (7.92e-5)</td>
      <td>unprognostic (1.39e-1)</td>
      <td>unprognostic (8.12e-2)</td>
      <td>unprognostic (1.95e-1)</td>
      <td>unprognostic (6.72e-2)</td>
    </tr>
    <tr>
      <th>20160</th>
      <td>ZZEF1</td>
      <td>FLJ10821, KIAA0399, ZZZ4</td>
      <td>ENSG00000074755</td>
      <td>Zinc finger ZZ-type and EF-hand domain contain...</td>
      <td>O43149</td>
      <td>17</td>
      <td>4004445-4143030</td>
      <td>Predicted membrane proteins</td>
      <td>Transcription, Transcription regulation</td>
      <td>Activator</td>
      <td>...</td>
      <td>unprognostic (1.44e-2)</td>
      <td>unprognostic (1.23e-1)</td>
      <td>unprognostic (2.21e-2)</td>
      <td>prognostic favorable (6.08e-4)</td>
      <td>unprognostic (1.54e-1)</td>
      <td>unprognostic (1.38e-3)</td>
      <td>unprognostic (6.09e-3)</td>
      <td>unprognostic (1.80e-1)</td>
      <td>unprognostic (9.43e-2)</td>
      <td>unprognostic (6.46e-2)</td>
    </tr>
    <tr>
      <th>20161</th>
      <td>ZZZ3</td>
      <td>ATAC1, DKFZP564I052</td>
      <td>ENSG00000036549</td>
      <td>Zinc finger ZZ-type containing 3</td>
      <td>Q8IYH5</td>
      <td>1</td>
      <td>77562416-77683419</td>
      <td>Predicted intracellular proteins</td>
      <td>Transcription, Transcription regulation</td>
      <td>DNA-binding</td>
      <td>...</td>
      <td>unprognostic (2.91e-1)</td>
      <td>unprognostic (9.67e-2)</td>
      <td>unprognostic (1.35e-1)</td>
      <td>unprognostic (2.83e-1)</td>
      <td>unprognostic (1.91e-1)</td>
      <td>unprognostic (1.71e-1)</td>
      <td>unprognostic (3.96e-1)</td>
      <td>unprognostic (1.95e-1)</td>
      <td>unprognostic (3.05e-2)</td>
      <td>unprognostic (1.43e-1)</td>
    </tr>
  </tbody>
</table>
<p>20162 rows × 89 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_hpa_expression_profile</span><span class="p">(</span><span class="n">protein</span><span class="p">):</span>
    <span class="n">HPA</span> <span class="o">=</span> <span class="n">HPA_dataset</span><span class="p">[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;RNA cell line specific nTPM&#39;</span><span class="p">]]</span>
    <span class="n">HPA</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">HPA</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span> <span class="s2">&quot;RNA cell line specific nTPM&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</section>
<section id="CTSE-overexpression-in-Gastric,-pancreatic,-and-colorectal-cancer-of-HPA-dataset">
<h3>CTSE overexpression in Gastric, pancreatic, and colorectal cancer of HPA dataset<a class="headerlink" href="#CTSE-overexpression-in-Gastric,-pancreatic,-and-colorectal-cancer-of-HPA-dataset" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">get_hpa_expression_profile</span><span class="p">(</span><span class="s2">&quot;CTSE&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;colorectal cancer: 28.4;Esophageal cancer: 48.8;Gastric cancer: 110.7;pancreatic cancer: 35.3&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HPA_protein_CTSE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Bladder Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Colorectal Carcinoma&#39;</span><span class="p">:</span> <span class="mf">28.4</span><span class="p">,</span>
    <span class="s1">&#39;Gastric Carcinoma&#39;</span><span class="p">:</span> <span class="mf">110.7</span><span class="p">,</span>
    <span class="s1">&#39;Glioma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Hepatocellular Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Kidney Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;NSCLC&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Pancreatic Carcinoma&#39;</span><span class="p">:</span> <span class="mf">35.3</span><span class="p">,</span>
    <span class="s1">&#39;SCLC&#39;</span><span class="p">:</span> <span class="mi">0</span>
<span class="p">}</span>


<span class="n">HPA_protein_CTSE</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">HPA_protein_CTSE</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CTSE&#39;</span><span class="p">])</span>
<span class="n">HPA_protein_CTSE</span> <span class="o">=</span> <span class="n">HPA_protein_CTSE</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CTSE&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">HPA_protein_CTSE</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CTSE</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>110.7</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>35.3</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>28.4</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>NSCLC</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>SCLC</th>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">STAVER</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/model-evlauate/STAVER_data.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">raw_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;~/STAVER-revised/model-evlauate/raw_data.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">process_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">target_gene</span><span class="p">):</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span><span class="n">target_gene</span><span class="p">,</span> <span class="s1">&#39;Cancer_type&#39;</span><span class="p">]]</span>
    <span class="n">df</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Other Solid Carcinomas&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Cancer_type&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="n">target_gene</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rawdata_CTSE</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="s2">&quot;CTSE&quot;</span><span class="p">)</span>
<span class="n">rawdata_CTSE</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CTSE&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">rawdata_CTSE</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CTSE</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>13.064867</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>13.046290</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>12.683957</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>11.875120</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>11.738225</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>11.182161</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>10.270660</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>10.095740</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">STAVER_CTSE</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">STAVER</span><span class="p">,</span> <span class="s2">&quot;CTSE&quot;</span><span class="p">)</span>
<span class="n">STAVER_CTSE</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CTSE&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">STAVER_CTSE</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CTSE</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>9.501038</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>8.500000</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>8.050000</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;pdf.fonttype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">42</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span><span class="o">=</span> <span class="mi">150</span> <span class="c1"># 一般将dpi设置在150到300之间</span>

<span class="n">custom_params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;axes.spines.right&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;axes.spines.top&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_theme</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;ticks&quot;</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="n">custom_params</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_barplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">target_col</span><span class="p">,</span> <span class="n">log_transform</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">x_label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">y_label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a barplot.</span>

<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): The dataframe.</span>
<span class="sd">        target_col (str): The column name to be plotted as the target column.</span>
<span class="sd">        x_label (str): The x-axis label. Default is None.</span>
<span class="sd">        y_label (str): The y-axis label. Default is None.</span>
<span class="sd">        title (str): The title of the plot. Default is None.</span>
<span class="sd">        save_path (str): The path to save the image. Default is None, indicating no saving.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.7</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">log_transform</span><span class="p">:</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">target_col</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="c1"># Use seaborn&#39;s barplot function to plot the graph</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">target_col</span><span class="p">])</span>

    <span class="c1"># Set the title and axis labels</span>
    <span class="k">if</span> <span class="n">title</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">x_label</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="n">x_label</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Index&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">y_label</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">y_label</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">target_col</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>  <span class="c1"># Rotate the x-axis labels for better display</span>

    <span class="c1"># Save the image (if save_path is specified)</span>
    <span class="k">if</span> <span class="n">save_path</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

    <span class="c1"># Show the image</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># 全局的绘图参数设置</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;pdf.fonttype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">42</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">150</span>

<span class="n">custom_params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;axes.spines.right&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;axes.spines.top&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_theme</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;ticks&quot;</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="n">custom_params</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_combined_barplots</span><span class="p">(</span><span class="n">datasets</span><span class="p">,</span> <span class="n">target_col</span><span class="p">,</span> <span class="n">titles</span><span class="p">,</span> <span class="n">log_transforms</span><span class="p">,</span> <span class="n">x_labels</span><span class="p">,</span> <span class="n">y_labels</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots multiple barplots on a single figure, one for each dataset provided.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lst</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">datasets</span><span class="p">)</span> <span class="k">for</span> <span class="n">lst</span> <span class="ow">in</span> <span class="p">[</span><span class="n">titles</span><span class="p">,</span> <span class="n">log_transforms</span><span class="p">,</span> <span class="n">x_labels</span><span class="p">,</span> <span class="n">y_labels</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;All list arguments must be of the same length as the &#39;datasets&#39; list.&quot;</span><span class="p">)</span>

    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">datasets</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="mi">4</span><span class="p">,</span> <span class="mf">4.2</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># If there is only one dataset, axes will not be an array</span>
        <span class="n">axes</span> <span class="o">=</span> <span class="p">[</span><span class="n">axes</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span> <span class="n">log_transform</span><span class="p">,</span> <span class="n">x_label</span><span class="p">,</span> <span class="n">y_label</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">datasets</span><span class="p">,</span> <span class="n">titles</span><span class="p">,</span> <span class="n">log_transforms</span><span class="p">,</span> <span class="n">x_labels</span><span class="p">,</span> <span class="n">y_labels</span><span class="p">)):</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">plot_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">target_col</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">log_transform</span> <span class="k">else</span> <span class="n">df</span><span class="p">[</span><span class="n">target_col</span><span class="p">]</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">plot_data</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">x_label</span> <span class="k">if</span> <span class="n">x_label</span> <span class="k">else</span> <span class="s2">&quot;Index&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">y_label</span> <span class="k">if</span> <span class="n">y_label</span> <span class="k">else</span> <span class="n">target_col</span><span class="p">)</span>

        <span class="c1"># Set the tick parameters and label rotation</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>

    <span class="c1"># Adjust the left margin if needed</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># You might need to tweak this value</span>

    <span class="k">if</span> <span class="n">save_path</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


<span class="c1"># The barplot of CTSE expression profiles</span>
<span class="n">plot_combined_barplots</span><span class="p">(</span>
    <span class="n">datasets</span><span class="o">=</span><span class="p">[</span><span class="n">HPA_protein_CTSE</span><span class="p">,</span> <span class="n">rawdata_CTSE</span><span class="p">,</span> <span class="n">STAVER_CTSE</span><span class="p">],</span>
    <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;CTSE&quot;</span><span class="p">,</span>
    <span class="n">titles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Expression profiles from HPA dataset (CTSE)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Expression profiles from original data (CTSE)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Expression profiles from STAVER data (CTSE)&quot;</span><span class="p">],</span>
    <span class="n">log_transforms</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
    <span class="n">x_labels</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
    <span class="n">y_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">],</span>
    <span class="n">save_path</span><span class="o">=</span><span class="s2">&quot;combined_barplots_CTSE.pdf&quot;</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_53_0.png" src="_images/PanCancer_949_celline_analysis_53_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The barplot of CTSE expression profiles</span>
<span class="n">plot_combined_barplots</span><span class="p">(</span>
        <span class="n">datasets</span><span class="o">=</span><span class="p">[</span><span class="n">HPA_protein_CTSE</span><span class="p">,</span> <span class="n">rawdata_CTSE</span><span class="p">,</span> <span class="n">STAVER_CTSE</span><span class="p">],</span>
        <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;CTSE&quot;</span><span class="p">,</span>
        <span class="n">titles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Expression profiles from HPA dataset (CTSE)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from original data (CTSE)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from STAVER data (CTSE)&quot;</span><span class="p">],</span>
        <span class="n">log_transforms</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
        <span class="n">x_labels</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
        <span class="n">y_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">],</span>
        <span class="n">save_path</span><span class="o">=</span><span class="s2">&quot;combined_barplots_CTSE.pdf&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="GPA33-overexpression-in-colorectal-cancer-and-Gastric-cancer-of-HPA-dataset">
<h3>GPA33 overexpression in colorectal cancer and Gastric cancer of HPA dataset<a class="headerlink" href="#GPA33-overexpression-in-colorectal-cancer-and-Gastric-cancer-of-HPA-dataset" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">get_hpa_expression_profile</span><span class="p">(</span><span class="s2">&quot;GPA33&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;Bile duct cancer: 34.1;colorectal cancer: 100.5;Gastric cancer: 19.1&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HPA_protein_CTSE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Bladder Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Colorectal Carcinoma&#39;</span><span class="p">:</span> <span class="mf">100.5</span><span class="p">,</span>
    <span class="s1">&#39;Gastric Carcinoma&#39;</span><span class="p">:</span> <span class="mf">19.1</span><span class="p">,</span>
    <span class="s1">&#39;Glioma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Hepatocellular Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Kidney Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;NSCLC&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Pancreatic Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;SCLC&#39;</span><span class="p">:</span> <span class="mi">0</span>
<span class="p">}</span>


<span class="n">HPA_protein_GPA33</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">HPA_protein_CTSE</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GPA33&#39;</span><span class="p">])</span>
<span class="n">HPA_protein_GPA33</span> <span class="o">=</span> <span class="n">HPA_protein_GPA33</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GPA33&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">HPA_protein_GPA33</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GPA33</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>100.5</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>19.1</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>NSCLC</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>SCLC</th>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rawdata_GPA33</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="s2">&quot;GPA33&quot;</span><span class="p">)</span>
<span class="n">rawdata_GPA33</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GPA33&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">rawdata_GPA33</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GPA33</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>16.132530</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>14.659133</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>13.211101</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>10.548243</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>9.851693</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>9.565263</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>9.219547</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>9.125549</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>9.036120</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">STAVER_GPA33</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">STAVER</span><span class="p">,</span> <span class="s2">&quot;GPA33&quot;</span><span class="p">)</span>
<span class="n">STAVER_GPA33</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GPA33&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">STAVER_GPA33</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GPA33</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>9.632096</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>7.820000</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The barplot of GPA33 expression profiles</span>
<span class="n">plot_combined_barplots</span><span class="p">(</span>
        <span class="n">datasets</span><span class="o">=</span><span class="p">[</span><span class="n">HPA_protein_GPA33</span><span class="p">,</span> <span class="n">rawdata_GPA33</span><span class="p">,</span> <span class="n">STAVER_GPA33</span><span class="p">],</span>
        <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;GPA33&quot;</span><span class="p">,</span>
        <span class="n">titles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Expression profiles from HPA dataset (GPA33)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from original data (GPA33)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from STAVER data (GPA33)&quot;</span><span class="p">],</span>
        <span class="n">log_transforms</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
        <span class="n">x_labels</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
        <span class="n">y_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">],</span>
        <span class="n">save_path</span><span class="o">=</span><span class="s2">&quot;combined_barplots_GPA33.pdf&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_60_0.png" src="_images/PanCancer_949_celline_analysis_60_0.png" />
</div>
</div>
</section>
<section id="ADGRF1-overexpression-in-pancreatic-cancer-of-HPA-dataset">
<h3>ADGRF1 overexpression in pancreatic cancer of HPA dataset<a class="headerlink" href="#ADGRF1-overexpression-in-pancreatic-cancer-of-HPA-dataset" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">get_hpa_expression_profile</span><span class="p">(</span><span class="s2">&quot;ADGRF1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;pancreatic cancer: 62.2&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HPA_protein_CTSE</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Bladder Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Colorectal Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Gastric Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Glioma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Hepatocellular Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Kidney Carcinoma&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;NSCLC&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s1">&#39;Pancreatic Carcinoma&#39;</span><span class="p">:</span> <span class="mf">62.2</span><span class="p">,</span>
    <span class="s1">&#39;SCLC&#39;</span><span class="p">:</span> <span class="mi">0</span>
<span class="p">}</span>


<span class="n">HPA_protein_ADGRF1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">HPA_protein_CTSE</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ADGRF1&#39;</span><span class="p">])</span>
<span class="n">HPA_protein_ADGRF1</span> <span class="o">=</span> <span class="n">HPA_protein_ADGRF1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ADGRF1&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">HPA_protein_ADGRF1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ADGRF1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>62.2</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>NSCLC</th>
      <td>0.0</td>
    </tr>
    <tr>
      <th>SCLC</th>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rawdata_ADGRF1</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="s2">&quot;ADGRF1&quot;</span><span class="p">)</span>
<span class="n">rawdata_ADGRF1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ADGRF1&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">rawdata_ADGRF1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ADGRF1</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>10.272240</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>10.256628</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>10.093484</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>9.910033</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">STAVER_ADGRF1</span> <span class="o">=</span> <span class="n">process_data</span><span class="p">(</span><span class="n">STAVER</span><span class="p">,</span> <span class="s2">&quot;ADGRF1&quot;</span><span class="p">)</span>
<span class="n">STAVER_ADGRF1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ADGRF1&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">STAVER_ADGRF1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ADGRF1</th>
    </tr>
    <tr>
      <th>Cancer_type</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Pancreatic Carcinoma</th>
      <td>7.594326</td>
    </tr>
    <tr>
      <th>Bladder Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Colorectal Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Gastric Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Glioma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Hepatocellular Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Kidney Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Non-Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Small Cell Lung Carcinoma</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The barplot of ADGRF1 expression profiles</span>
<span class="n">plot_combined_barplots</span><span class="p">(</span>
        <span class="n">datasets</span><span class="o">=</span><span class="p">[</span><span class="n">HPA_protein_ADGRF1</span><span class="p">,</span> <span class="n">rawdata_ADGRF1</span><span class="p">,</span> <span class="n">STAVER_ADGRF1</span><span class="p">],</span>
        <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;ADGRF1&quot;</span><span class="p">,</span>
        <span class="n">titles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Expression profiles from HPA dataset (ADGRF1)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from original data (ADGRF1)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Expression profiles from STAVER data (ADGRF1)&quot;</span><span class="p">],</span>
        <span class="n">log_transforms</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
        <span class="n">x_labels</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
        <span class="n">y_labels</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Log2 (Relative expression)&quot;</span><span class="p">],</span>
        <span class="n">save_path</span><span class="o">=</span><span class="s2">&quot;combined_barplots_ADGRF1.pdf&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/PanCancer_949_celline_analysis_66_0.png" src="_images/PanCancer_949_celline_analysis_66_0.png" />
</div>
</div>
</section>
</section>
<section id="The-reproducibility-of-the-STAVER-algorithm-in-identifying-the-previously-reported-tumor-biomarkers">
<h2>The reproducibility of the STAVER algorithm in identifying the previously reported tumor biomarkers<a class="headerlink" href="#The-reproducibility-of-the-STAVER-algorithm-in-identifying-the-previously-reported-tumor-biomarkers" title="Link to this heading">¶</a></h2>
<p>To further validate the superior reproducibility of the STAVER algorithm in identifying tumor biomarkers, a systematic literature review was conducted. According to the systematic literature review, a series of cancer biomarkers were selected to further evaluate the STAVER algorithm <strong>(Table RL1 and Table RL4)</strong>. We utilized the original and STAVER-processed proteomic data to examine the expression differences of these previously reported cancer biomarkers across diverse cancer cell lines.</p>
<p>The results demonstrated that the STAVER algorithm more accurately identify the previously reported tumor biomarkers with high reproducibility.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reported_biomrkers</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_clipboard</span><span class="p">()</span>
<span class="n">reported_biomrkers</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Tissue Type</th>
      <th>Cancer Type</th>
      <th>Reported Biomarker</th>
      <th>Reference</th>
      <th>PMID</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Liver</td>
      <td>Hepatocellular carcinoma</td>
      <td>DCP</td>
      <td>Journal of Hepatology, 2023</td>
      <td>PMID: 37683735</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Liver</td>
      <td>Hepatocellular carcinoma</td>
      <td>HSP70</td>
      <td>Journal of hepatology, 2009</td>
      <td>PMID: 19231003</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Liver</td>
      <td>Hepatocellular carcinoma</td>
      <td>GGT1</td>
      <td>Gastroenterology, 2017; BMC Cancer, 2019</td>
      <td>PMID: 28711626;  PMID: 31455253</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Liver</td>
      <td>Hepatocellular carcinoma</td>
      <td>A1CF</td>
      <td>Cell reports, 2019; The JCI, 2021</td>
      <td>PMID: 31597092; PMID: 33445170</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Kidney</td>
      <td>Kidney cancer</td>
      <td>CA9</td>
      <td>European urology, 2014; Cancer research, 1997</td>
      <td>PMID: 24821582; PMID: 9230182</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Kidney</td>
      <td>Kidney cancer</td>
      <td>CD70</td>
      <td>Cancer research, 2006</td>
      <td>PMID: 16489038</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Gastric</td>
      <td>Gastric cancer</td>
      <td>ERRB2</td>
      <td>Annals of oncology, 2008</td>
      <td>PMID: 18441328</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Gastric</td>
      <td>Gastric cancer</td>
      <td>AGR3</td>
      <td>In vivo, 2023</td>
      <td>PMID: 36593009</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Colorectal</td>
      <td>Colorectal cancer</td>
      <td>MUC2</td>
      <td>Gastroenterology, 2005;</td>
      <td>PMID: 16285957</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Colorectal</td>
      <td>Colorectal cancer</td>
      <td>EPCAM</td>
      <td>British journal of cancer, 2014</td>
      <td>PMID: 24786601</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Colorectal</td>
      <td>Colorectal cancer</td>
      <td>CDX2</td>
      <td>Annals of oncology, 2017</td>
      <td>PMID: 28328000</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Colorectal</td>
      <td>Colorectal cancer</td>
      <td>MUC13</td>
      <td>Oncogene, 2019</td>
      <td>PMID: 31427737</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Brain</td>
      <td>Glioma</td>
      <td>GFAP</td>
      <td>Brain, 2007; Biosens Bioelectron, 2020</td>
      <td>PMID: 17998256; PMID: 33160234</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Brain</td>
      <td>Glioma</td>
      <td>NES</td>
      <td>Cancer cell, 2020; Mol Neurobiol, 2017</td>
      <td>PMID: 32396858; PMID: 26768429</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Pancreas</td>
      <td>Pancreatic cancer</td>
      <td>MUC1</td>
      <td>Journal of gastroenterology,2003</td>
      <td>PMID: 14714254</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Pancreas</td>
      <td>Pancreatic cancer</td>
      <td>ANO1</td>
      <td>PNAS, 2019</td>
      <td>PMID: 31182586</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Lung</td>
      <td>Non-small cell lung cancer</td>
      <td>KRT5</td>
      <td>Cancer cell, 2022</td>
      <td>PMID: 36368318</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Lung</td>
      <td>Non-small cell lung cancer</td>
      <td>DSC3</td>
      <td>Journal of thoracic oncology, 2011</td>
      <td>PMID: 21623236</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Lung</td>
      <td>Small cell lung cancer</td>
      <td>CHGA</td>
      <td>Cell reports, 2020</td>
      <td>PMID: 33086069</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Lung</td>
      <td>Small cell lung cancer</td>
      <td>NCAM1</td>
      <td>Cancer cell, 2022</td>
      <td>PMID: 36368318</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<img alt="Reported_cancer_biomarkers" src="%22/Users/ranpeng/Downloads/reported_cancer_biomarkers.png%22" />
<p>All the results are shown in the above figure and can be reproduced with the R script <code class="docutils literal notranslate"><span class="pre">boxplot_949_PanCancer.R</span></code>.</p>
</section>
<section id="The-robustness-and-broad-applicability-of-the-STAVER-algorithm-for-disease-diagnosis-and-classification.">
<h2>The robustness and broad applicability of the STAVER algorithm for disease diagnosis and classification.<a class="headerlink" href="#The-robustness-and-broad-applicability-of-the-STAVER-algorithm-for-disease-diagnosis-and-classification." title="Link to this heading">¶</a></h2>
<p>To validate the outstanding advantages of the STAVER algorithm for potential clinical decision-making, we comprehensively evaluated the generalization performance and prediction accuracy of classification models constructed based on the above tumor biomarkers.</p>
<p>To illustrate the robustness and broad applicability of the STAVER algorithm, we constructed three separate benchmark classification models based on different algorithms, including decision tree-based random forest, gradient boosting-based XGBoost, and linear regression-based logistic regression models (Materials and methods section).</p>
<p>As a result, compared to the original proteomic data, the classification models constructed based on STAVER-processed data exhibited increased generalization performance in distinguishing specific cancer-type cell lines.</p>
<p>See the <a class="reference external" href="https://doi:10.1016/j.ccell.2022.06.010">PanCancer_949_cellline_models.ipynb</a> for more details.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="staver_pipeline_cli.html" class="btn btn-neutral float-left" title="Run the STAVER workflow" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="PanCancer_949_cellline_models.html" class="btn btn-neutral float-right" title="The robustness and broad applicability of the STAVER algorithm for disease diagnosis and classification." accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab, 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>